### PR TITLE
GDP CP Support

### PR DESCRIPTION
**Pull request by Mellonta**
## Summary
This PR adds chunkwise context parallelism for the FLA implementation of Gated Delta Product (GDP).
The main change is a unified chunkwise CP interface that separates Megatron’s distributed orchestration from backend-specific linear attention kernels.

## Interface
LinearAttentionCPBackend divides the operation into four stages:
- cp_forward_prepare: compute the local recurrent-state summary with a zero incoming state.
- cp_forward_apply: combine preceding-rank summaries and compute the local output.
- cp_backward_prepare: compute the local backward summary with a zero outgoing-state gradient.
- cp_backward_apply: combine following-rank summaries and finish the local input gradients.

Megatron owns:

- CP process groups and collectives.
- All-gathering local summaries.
- Selecting the causally relevant prefix or suffix of gathered summaries.

The backend doesn't see process groups

## FLA GDP backend

The GDP kernel implementation is mostly adapted from FLA’s chunkwise context-parallel implementation.
The Megatron-specific code adapts those kernels to the new CP interface.

## Integration
- Adds --linear-cp-mode chunkwise and --linear-cp-layout contiguous.

## Tests
```
uv run python -m torch.distributed.run --nproc-per-node 4 -m pytest -q \
    tests/unit_tests/ssm/test_chunkwise_context_parallel.py \
    tests/unit_tests/ssm/test_gdp_chunkwise_context_parallel.py \
    tests/unit_tests/ssm/test_gdp_packed_seq.py
```

Validated on Nemotron models

### CODE REVIEW

**Comment by ZhiyuLi-Nvidia** [comment]
Hi, @Mellonta thank you for the awesome feature.

Does the feature support deterministic mode?

Let me know if there's additional steps for determinism specifically. We can work together.

cc @Connor-XY @ericharper

**Review by zhongbozhu — APPROVED** [review]

File: megatron/core/ssm/gated_delta_product.py [hunk 1]
```diff
@@ -513,29 +551,26 @@ def forward(
...
-        points. Both are ``None`` when sequences are not packed.
-        """
+    def _packed_metadata(self, packed_seq_params):
+        """Return sequence indices and cumulative lengths for packed input."""
         if packed_seq_params is None:
             return None, None
-        return build_packed_seq_idx(packed_seq_params, pack_length), get_cu_seqlens(
-            packed_seq_params
-        )
+        return packed_seq_params.seq_idx, get_cu_seqlens(packed_seq_params)
```
[the remark this answers is not in the document]
Comment (Mellonta) [reply megatron/core/ssm/gated_delta_product.py:576]: Intentional, because old code may cause divergence.

File: megatron/core/models/hybrid/hybrid_block.py [hunk 2]
```diff
@@ -418,12 +441,24 @@ def get_inner_quant_context(config, layer_number):
                                 padding_mask=padding_mask,
                             )
                         else:  # MambaLayer, Expert, or MLP
-                            hidden_states = layer(
-                                hidden_states=hidden_states,
-                                attention_mask=attention_mask,
-                                inference_context=inference_context,
-                                packed_seq_params=layer_packed_seq_params,
-                            )
+                            if (
```
Comment (Phlip79) [inline megatron/core/models/hybrid/hybrid_block.py]: Make this an `elif`

**Comment by shcho1118** [comment]
Hi @Mellonta, thanks for the great work on GDP chunkwise CP!

I recently opened #6808 to add state-passing context parallelism for Mamba2 SSD. Both PRs share a very similar design philosophy: decomposing the recurrent state into local summaries (`cp_forward_prepare` / `S_ext, a_block`) and performing an all-gather/scan across CP ranks to avoid full activation redistribution.

I'd love to check your thoughts on:

1. **Interface Unification:** Do you envision `LinearAttentionCPBackend` (`cp_forward_prepare/apply`, `cp_backward_prepare/apply`) becoming the unified standard across all linear attention and SSM layers (including Mamba2 SSD)?
2. **Layout & Load Balancing:** PR #6815 currently enforces `linear_cp_layout='contiguous'`. In #6808, we explored a `virtual` mode that keeps the zigzag/balanced layout directly without permutation. Would you be open to aligning the layout handling or sharing the boundary collective orchestration between these two paths?

Curious to hear your thoughts on aligning the two approaches.

**Comment by Mellonta** [comment]
Hi @shcho1118 , regarding your questions:

1. Yes, this is a unified interface for all linear attention variants. We also have a Mamba2 implementation using this interface, but haven't published it yet.
2. We do not plan to support a zigzag layout in linear attention layers because it hurts kernel efficiency and doubles the communication required for state passing.

**Comment by shcho1118** [comment]
> 2\. We do not plan to support a zigzag layout in linear attention layers because it hurts kernel efficiency and doubles the communication required for state passing.

Thanks for clarifying! Regarding the concerns around kernel efficiency and communication volume in virtual mode, looking at the actual computation and communication trade-offs shows why virtual layout is advantageous:

1. **Computation Overhead & Kernel Efficiency**
   - **Overhead is strictly isolated to boundary stages:** The 2x batch expansion overhead only applies to the boundary summary stages (`cp_*_prepare` / `cp_*_apply`), which account for a small fraction of the total layer execution time.
   - **Core chunkwise kernels show zero efficiency penalty:** For the main chunkwise computation kernels (e.g., intra-chunk linear attention GEMMs, SSD chunk scan), total FLOPs and memory access remain strictly identical (2 * (L / 2) = L). In our kernel benchmarks on B200, packing the segments onto the batch axis introduces no kernel efficiency degradation.

2. **Communication: State Summary vs. Activation Permutation**
   - While state passing communication does double (2 * CP summary entries instead of CP), this payload is fixed and sequence-length-independent (only a few MiB of FP32 boundary states).
   - In contrast, supporting `contiguous` layout inside a standard Megatron zigzag-balanced CP requires physical activation redistribution (`zigzag` <-> `contiguous` permutation), which scales with O(sequence length).
   - Communicating a slightly larger state summary (O(1) w.r.t. sequence length) is orders of magnitude cheaper than the full activation exchange required to make shards contiguous.

In our end-to-end benchmarks, avoiding activation permutations via virtual mode consistently yielded ~1.5x–2.0x speedups over permute-based contiguous approaches (and A2A).

**Comment by Mellonta** [comment]
Hi @shcho1118, 
We tested both approaches at scale but observed no speedups in our workloads. Most performance factors remain constant, and the ratio of linear layers to softmax attention plays a major role. I agree certain configurations likely benefit from this virtual setup. The main question is whether those use cases are common enough to justify the added complexity in Megatron, given the need for universal backend support. If we can integrate them without excessive overhead across different backends, I believe it would be a valuable feature to support.

**Comment by shcho1118** [comment]
Hi @Mellonta 

Thanks for the detailed response!

I completely agree with the concern regarding codebase complexity in Megatron. Keeping the backend interface clean, minimal, and universally maintainable across backends is definitely a priority.

To share a bit of context on our side: we validated this setup at scale on a Nemotron-3 Nano workload using **TP=2, CP=4, EP=8 across 128 nodes (1,024x B200)**. Model convergence and loss curves were identical to standard All-to-All (A2A). On end-to-end throughput:
* At **32k sequence length**, we observed roughly a **~10% speedup**.
* At **128k sequence length**, the throughput gain reached **~30%**.
* As you noted, the ratio of linear to attention layers matters significantly, and at shorter sequence lengths (e.g., 16k or below), the difference would likely be negligible.

Because we saw clear benefits in this virtual setup at longer contexts, I wanted to share our observations. I really appreciate you taking the time to engage in such an in-depth discussion!

File: megatron/core/recompute.py [hunk 3]
```diff
@@ -120,7 +125,13 @@ def custom_forward(
                     else:  # MambaLayer (HybridStack `M` slot)
                         for k in ("context", "context_mask", "attention_bias", "padding_mask"):
                             layer_kwargs.pop(k, None)
```
Comment (Phlip79) [inline megatron/core/recompute.py:125]: How about just adding:

```
if packed_sequence_cp_metadata is not None:
  layer_kwargs["packed_sequence_cp_metadata"] = packed_sequence_cp_metadata
```

File: megatron/core/models/hybrid/layers/hybrid_hyper_connection.py [hunk 4]
```diff
@@ -69,12 +71,21 @@ def _call_inner_layer(
             )
         else:
             # Mamba-like layers only consume the common HybridStack arguments.
-            output = self.inner_layer(
-                hidden_states=hidden_states,
-                attention_mask=attention_mask,
-                inference_context=inference_context,
-                packed_seq_params=packed_seq_params,
-            )
+            if packed_sequence_cp_metadata is None:
```
Comment (jaredcasper) [inline megatron/core/models/hybrid/layers/hybrid_hyper_connection.py]: Would it be cleaner to make an extra_kw_args dict here instead of splitting into two calls?

Or do you even need two calls here if packed_sequence_cp_metadata is None, can't you just pass it unconditionally?

File: megatron/core/models/hybrid/hybrid_block.py [hunk 5]
```diff
@@ -131,6 +135,9 @@ def __init__(
             "--hybrid-layer-pattern by HybridModel."
         )
         self.layer_type_list = layer_type_list
+        self._has_linear_layer_with_chunkwise_cp = (
+            self._linear_cp_is_chunkwise and LayerSymbols.MAMBA in self.layer_type_list
```
Comment (jaredcasper) [inline megatron/core/models/hybrid/hybrid_block.py]: @Phlip79 what would this look like in your new config scheme?

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/models/hybrid/layers/hybrid_hyper_connection.py (hunk 4 above)
Comment (Mellonta) [reply megatron/core/models/hybrid/layers/hybrid_hyper_connection.py]: No, we couldn't pass it unconditionally because the other path doesn't have this argument. But I can use extra_kw_args to make it cleaner.

File: megatron/core/models/hybrid/hybrid_block.py (hunk 5 above)
Comment (Phlip79) [reply megatron/core/models/hybrid/hybrid_block.py]: We would want to add a new function `contains_layer_symbol` that takes a `LayerSymbols` and `layer_config_list`.

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]
